stdlib/public/Cxx/CxxConvertibleToCollection.swift

```diff
@@ -21,36 +21,36 @@ public protocol CxxConvertibleToCollection {
   mutating func __endUnsafe() -> RawIterator
 }
 
-@inlinable @inline(__always)
-internal func forEachElement<C: CxxConvertibleToCollection>(
-  of c: C,
-  body: (C.RawIterator.Pointee) -> Void
-) {
-  var mutableC = c
-  withExtendedLifetime(mutableC) {
-    var rawIterator = mutableC.__beginUnsafe()
-    let endIterator = mutableC.__endUnsafe()
-    while rawIterator != endIterator {
-      body(rawIterator.pointee)
-      rawIterator = rawIterator.successor()
+extension CxxConvertibleToCollection {
+  @inlinable
+  internal func forEach(_ body: (RawIterator.Pointee) -> Void) {
+    var mutableSelf = self
+    withExtendedLifetime(mutableSelf) {
+      var rawIterator = mutableSelf.__beginUnsafe()
+      let endIterator = mutableSelf.__endUnsafe()
+      while rawIterator != endIterator {
+        body(rawIterator.pointee)
+        rawIterator = rawIterator.successor()
+      }
     }
   }
 }
 
-extension Array {
-  /// Creates an array containing the elements of a C++ container.
+extension RangeReplaceableCollection {
+  /// Creates a collection containing the elements of a C++ container.
   ///
-  /// This initializes the array by copying every element of the C++ container.
+  /// This initializes the collection by copying every element of the C++
+  /// container.
   ///
   /// - Complexity: O(*n*), where *n* is the number of elements in the C++
   ///   container when each element is copied in O(1). Note that this might not
   ///   be true for certain C++ types, e.g. those with a custom copy
   ///   constructor that performs additional logic.
-  public init<C: CxxConvertibleToCollection>(_ c: C)
+  public init<C: CxxConvertibleToCollection>(contentsOf c: C)
     where C.RawIterator.Pointee == Element {
 
     self.init()
-    forEachElement(of: c) { self.append($0) }
+    c.forEach { self.append($0) }
```

Comment:
Did we talk about the idea of reserving capacity here?

```swift
self.init()
reserveCapacity(c.count)
c.forEach { self.append($0) }
```

This will only be beneficial to _some_ range-replaceable collection types, and generally only if we can rely on `c.count` to be fast (faster than iteration), but when it works, preallocating storage up front can roughly double overall throughput.

Reply:
Unfortunately we don't have a reliable way to determine the number of elements in a C++ container. Some containers from the stdlib provide a `size()` method, but custom user containers might not provide API for `count`.